src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,323 @@
+		fallthrough
+	case syntax.OpConcat:
+		// strip left-most '^'
+		if l := len(parsed.Sub); leftmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost && l == 1)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+		// strip right-most '$'
+		if l := len(parsed.Sub); rightmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[l-1], leftmost && l == 1, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[l-1] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpAlternate:
+		// strip left-most '^' and right-most '$' in each sub-expression
+		for idx := range parsed.Sub {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[idx], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			parsed.Sub[idx] = newRe
+			changed = changed || c
+		}
+	case syntax.OpQuest:
+		if len(parsed.Sub) > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpStar:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					&syntax.Regexp{
+						Op:    syntax.OpQuest,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							newRe,
+						},
+					},
+					original,
+				},
+			}
+			changed = true
+		}
+	case syntax.OpPlus:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			return &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					&syntax.Regexp{
+						Op:    syntax.OpStar,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							original.Sub[0],
+						},
+					},
+				},
+			}, true, nil
+		}
+	case syntax.OpRepeat:
+		if len(parsed.Sub) > 0 && parsed.Min > 0 && parsed.Max > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			original.Min--
+			original.Max--
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					newRe,
+					original,
+				},
+			}
+			changed = true
+		}
+	}
+	return newRegexp, changed, nil
```

Comment:
Would it make more sense to just return in each sub switch case expression and just `return input, false, nil` if we get to the bottom?  Seems none of them need any code after the case statements.

Reply:
Sure thing, that'll read better